kivy/core/image/__init__.py

```diff
@@ -626,28 +658,68 @@ def anim_reset(self, allow_anim):
             self._anim_ev.cancel()
             self._anim_ev = None
 
-        if allow_anim and self._anim_available and self._anim_delay >= 0:
-            self._anim_ev = Clock.schedule_interval(self._anim,
-                                                    self.anim_delay)
-            self._anim()
+        if allow_anim and self._anim_available:
+            if self._durations:
+                if not len(self._durations) == len(self.image.durations):
+                    raise Exception(
+                        'Image frames count not equal durations count')
+                # there are minimum 2 elements otherwise this is constant case
+                # and durations is None, this index is for the next frame
+                self._anim_index = 1
+                self._anim_ev = Clock.schedule_once(self._anim,
+                                                    self._durations[0])
+            elif self._anim_delay >= 0:
+                self._anim_ev = Clock.schedule_interval(self._anim,
+                                                        self._anim_delay)
 
     def _get_anim_delay(self):
         return self._anim_delay
 
     def _set_anim_delay(self, x):
-        if self._anim_delay == x:
+        if self._anim_delay == x or self._durations:
             return
         self._anim_delay = x
         if self._anim_available:
             if self._anim_ev is not None:
                 self._anim_ev.cancel()
                 self._anim_ev = None
 
-            if self._anim_delay >= 0:
+            # start animation only if durations is empty,
+            # otherwise start animation in _set_durations()
+            if self._anim_delay >= 0 and not self._durations:
                 self._anim_ev = Clock.schedule_interval(self._anim,
                                                         self._anim_delay)
 
     anim_delay = property(_get_anim_delay, _set_anim_delay)
+
+    def _get_durations(self):
+        return self._durations
+
+    def _set_durations(self, x):
+        # Waiting for list of ints or floats in microseconds like in GIF
+        # format, not seconds
+        if x is None:
+            self._durations = None
+            # if durations set None stop until anim_delay set again
+            self.anim_reset(False)
+            return
+        if x and not(isinstance(x, list)
+                     and all(isinstance(elem, (int, float)) for elem in x)):
+            raise Exception('Property durations must be list of int or '
+                            'float with len > 0')
+
+        if all(elem == x[0] for elem in x):
```

Comment:
There's really no reason to do this optimization, is there? It's simpler to trust the durations passed in. And if they are all the same the user would just have used `anim_delay` directly. And either way I don't think this would make it that much faster to be worth the added complexity of checking all the elements.

Reply:
**Durations from the GIF file get here too.** GIF-file with constant delay is more common than with variable. We check it once and use many times with common cases with constant delay. So, the real standart case is Image('image.gif'):
1) load GIF-file
2) GIF-file has constant delay(common case)
3) set this constant delay as anim_delay
4) create once Clock.schedule_interval()

Reply:
Additionally, if we can't simply overwrite `anim_delay` that may have been set by the user.

Reply:
But what is wrong with simply using the list of durations when they are all the same? I.e. even if it is all the same (i.e. `all(elem == x[0] for elem in x)` is True) why can't we do `self._durations = x`?

Is it possible that the length of x is not the same as the length of the frames? If that's what you're trying to check, then it's better to check for this directly!?

Reply:
If we want to consider anim_delay by default, we should set auto_anim_delay False. We must consider only one thing as dominant. We can't consoder durations as dominant and anim_delay as dominant together. I can rewrite code for one dominant or another but not for both. e.g. user set anim_delay and durations both what should we do? Combine anim_delay and durations in one property? I tried, but could not do it normally without errors.

Reply:
> But what is wrong with simply using the list of durations when they are all the same? I.e. even if it is all the same (i.e. `all(elem == x[0] for elem in x)` is True) why can't we do `self._durations = x`?
> 
> Is it possible that the length of x is not the same as the length of the frames? If that's what you're trying to check, then it's better to check for this directly!?

There is one reason. I think Clock.schedule_interval() once is faster then creating Clock.schedule_once() every frame.

Reply:
I may be misunderstanding something, but why can't we simply replace

```py
        if all(elem == x[0] for elem in x):
            # durations are constant, so we have standart anim_delay case
            self._durations = None
            self.anim_delay = x[0]  # anim resets inside _set_anim_delay()
                                    # if not durations
        else:
            # reset and stop for error case when durations count < frames count
            self.anim_reset(False)
            self._durations = x
            self.anim_reset(True)
```
with 
```py
# reset and stop for error case when durations count < frames count
self.anim_reset(False)
self._durations = x
self.anim_reset(True)
```

You mentioned about dominance, but isn't durations used first and then `anim_delay` if durations is empty? That's how it should be in my POV. But either way, if can be converted to a list successfully, we can just keep it as a list of durations, even if the durations are all the same values in the list!?

Reply:
I think there is reason not to change durations, ok. The user should expect the durations to remain in place. But I think we should still call once schedule_interval() for constant durations case just for performance.

Reply:
>There is one reason. I think Clock.schedule_interval() once is faster then creating Clock.schedule_once() every frame.

That's a good point, but as you mentioned, checking floats for equality is not exactly correct, so I think from a correctness POV it's better to be simply keep the list of durations rather than rounding and comparing values.

Also, I'd consider it a premature optimization. You're probably right it's a little faster, but in the grand scheme of things, drawing and copying the image probably dominates any speed issues so an additional `Clock.schedule_once()` once per frame is not likely to be noticeable. Especially since this should be quite small. Unless you have measured it.

Additionally, if you cared about performance and measured and found it is causing some lag, the better way is to create a single event with `event = Clock.create_event(...)` and then for each image manually set the timeout for the event and then cancel and trigger it with `event.cancel(); event()`. But all this is not really needed unless we have actual performance profiling data.

Reply:
Trust you, I don’t know as well what's going on under the hood as you. So I will do as you said: schedule_once every frame.